api/src/main/java/com/debugmafia/clueless/model/Board.java

```diff
@@ -0,0 +1,220 @@
+        }
+      }
+
+      winningCards = cardsToReturn;
+    } else {
+      cardsToReturn = winningCards;
+    }
+    return cardsToReturn;
+  }
+
+  public Set<Card> draw(int numToDraw) {
+
+    HashSet<Card> setToReturn = null;
+
+    if (hasDeckBeenShuffled && deckOCards.size() >= numToDraw) {
+
+      Iterator<Card> iterator = deckOCards.iterator();
+
+      while(numToDraw > 0 && iterator.hasNext())
+      {
+        setToReturn.add(iterator.next());
+        iterator.remove();
+        numToDraw--;
+      }
+    }
+    return setToReturn;
+  }
+
+  public Set<BoardLocation> getOpenAdjacentLocations(BoardLocation locationToQuery) {
+    Set<BoardLocation> validLocations = new HashSet<BoardLocation>();
+
+    for (BoardLocation[] row : grid) {
+      for(BoardLocation column: row)
+      {
+      if (locationToQuery.getXcoord() + 1 == column.getXcoord() || locationToQuery.getXcoord() + 1 == column.getYcoord()) {
+        if (column.getType() == BoardLocationType.ROOM) {
+          validLocations.add(column);
+        } else if (column.getType() == BoardLocationType.HALLWAY && locationToQuery.getPieces().isEmpty()) {
+          validLocations.add(column);
+        }
+      } else if (locationToQuery.getSecretPassage() != null && locationToQuery.getSecretPassage() == column) {
+        validLocations.add(column);
+      }
+    }
+    }
+
+    return validLocations;
+  }
+
+  public Board movePiece(Piece p, BoardLocation to) {
+
+    BoardLocation localLocation = null;
+    Piece localPiece = null;
+
+    for(BoardLocation[] row : grid )    
+    {
+      for(BoardLocation column : row)
+      {
+        if(column == to)
+        {
+          localLocation = column;
+          break;
+        }
+      }
+    }
+
+    for(Piece pIter : gamePieces)
+    {
+      if(pIter == p)
+      {
+        localPiece = pIter;
+        break;
+      }
+    }
+
+    if(localLocation == null || localPiece == null)
+    {
+      return null;
+    }
+
+    HashSet<Piece> localPieces = new HashSet<Piece>(localLocation.getPieces());
+
+    localPieces.add(localPiece);
+    localLocation.setPieces(localPieces);
+
+    return this;
+  }
+
+  public Board moveWeapons(Weapon w, BoardLocation to) {
+
+
+    BoardLocation localLocation = null;
+    Weapon localWeapon = null;
+
+    for(BoardLocation[] row : grid )    
+    {
+      for(BoardLocation column : row)
+      {
+        if(column == to)
+        {
+          localLocation = column;
+          break;
+        }
+      }
+    }
+
+    for(Weapon wIter : gameWeapons)
+    {
+      if(wIter == w)
+      {
+        localWeapon = wIter;
+        break;
+      }
+    }
+
+    if(localLocation == null || localWeapon == null)
+    {
+      return null;
+    }
+
+    HashSet<Weapon> localWeapons = new HashSet<Weapon>(localLocation.getWeapons());
+
+    localWeapons.add(localWeapon);
+    localLocation.setWeapons(localWeapons);
+
+    return this;
+
+  }
+
+  public Card getAssociatedCard(Weapon w) {
```

Comment:
Logic is sound here. I think we can improve it by:

1. Overwriting the `equals()` method in the `Weapon` class, putting the "equals" logic in there that you have here (type are equal, names are equal)
2. Updating this `filter` call to us the `.equals()` method, making this read much cleaner
3. Using the `findFirst` stream method that Java 8 comes with, that way you don't have to do a size check, and return the item at the 0th index

Reply:
Actually, I'm wrong here about the `.equals` stuff. This is grabbing a card from a weapon/piece/location....so what I said doesn't apply. Oops.